xblock/scorable.py

```diff
@@ -0,0 +1,131 @@
+"""
+Scorable.
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from collections import namedtuple
+import logging
+
+
+log = logging.getLogger(__name__)
+
+
+Score = namedtuple('Score', ['raw_earned', 'raw_possible'])
+
+
+class ScorableXBlockMixin(object):
+    """
+    Mixin to handle functionality related to scoring.
+
+    Provides the following public methods:
+
+        ScorableXBlockMixin.allows_rescore(self) -> bool
+            Returns True if the block can be rescored.
+
+    Subclasses must define the following:
+
+        ScorableXBlockMixin.get_score(self)
+            Return the currently saved score.  Must not calculate a new score.
+
+        ScorableXBlockMixin.set_score(self, score)
+            Set the score on the block to the specified value
+
+        ScorableXBlockMixin.calculate_score(self, score)
+            Calculate a new score based on the state of the block.  Must not
+            update any values on the block.
+    """
+
+    def rescore(self, only_if_higher):
+        """
+        Calculate a new raw score and save it to the block.  If only_if_higher
+        is True and the score didn't improve, keep the existing score.
+
+        Raises a TypeError if the block cannot be scored.
+        Raises a ValueError if the user has not yet completed the problem.
+        """
+
+        _ = self.runtime.service(self, 'i18n').ugettext
+        resultlog_prefix = _('Rescore result:')
+
+        if not self.allows_rescore():
+            error_msg = _('Problem does not support rescoring: {}').format(self.location)
+            raise TypeError(error_msg)
+
+        original_score = self.get_score()
+        if original_score is None:
+            error_msg = _('Cannot rescore unanswered problem: {}').format(self.location)
+            raise ValueError(error_msg)
+
+        new_score = self.calculate_score()
+
+        self.set_score(new_score)
+        self._publish_grade(only_if_higher)
+        if only_if_higher:
+            only_if_higher_msg = _('Rescore only if score improved.')
+        else:
+            only_if_higher_msg = _('Rescore unconditionally.')
+        result_msg = _('Original score: {}. New score: {}. {}').format(
+            original_score,
+            new_score,
+            only_if_higher_msg
+        )
+        log.info('{} {}'.format(resultlog_prefix, result_msg))
```

Comment:
Given that the comparison and setting of the scores is done in the grades layer, I suggest moving this log statement there - as that layer owns whether or not the rescore actually happens.
On the other hand, I don't see the need for this log statement.  The instructor task already emits an event for this: https://github.com/edx/edx-platform/blob/master/lms/djangoapps/instructor_task/tasks_helper.py#L580